### PR TITLE
[GH-4958] Platform behavior implementation

### LINKED ISSUE #4958: Proposal for new API to create Behaviors
### Description

With the multi-target arch. we don't need anymore the `Effects` API, created on Xamarin.Forms. This API is something like the `Behavior` API but when you need to access the native APIs to customize the control (e.g. TintColor or removing the border for the entry). Right now the `Behavior` API, from Xamarin.Forms, isn't suitable for this task, due to some instantiation order as you can see here #4496.

As discussed in #4496 the team will not change the actual behavior of Control initialization. With that, we should bring new alternatives. After some tries, I have one API that I named `BasePlatformBehavior` (naming is hard, feel free to change it). Before moving to it I would like to discuss the existing ways to customize a control without the `Effects` API. You can see all the projects [here](https://github.com/pictos/MauiSpecRepo).

> For this sample, I want to color a picture that is shown by an `Image` control.

## [Mapper](https://github.com/pictos/MauiSpecRepo/blob/master/MauiSpecRepo/Mapper/TintColorMapper.cs)

We can use the `ImageHandler.Mapper.Add` to add a new property and an `Action` to it.
`ImageHandler.Mapper.Add("TintColor", (handler, view) => { }).`
The `TintColor` is an Attached Property, so I can set it in any `Image` control and when that property changes that `Action` will trigger.

This is really straightforward to implement and use, but this lacks in some parts:
- This will be applied to all `Image` controls, if we have a lot of them (in other controls) can this impact the renderer time?
- Is not very to know when to clean up the customization, like here when I set `null` I know that I don't want to apply the tintColor, but may have other scenarios that `null` means something different.

## [Custom Control](https://github.com/pictos/MauiSpecRepo/blob/master/MauiSpecRepo/CustomControl/ImageTintColor.cs)

Thanks to the multi-target arch. I can access the native code on my shared control implementation. The way that I would implement the TintColor now is by using a normal `BindableProperty` and controlling it on my `PropertyChanged` method.

But unfortunately, there's some downsides, as you can see [here](https://github.com/pictos/MauiSpecRepo/blob/master/MauiSpecRepo/CustomControl/ImageTintColor.cs#L18-L26)
```csharp
        if (control.Handler is null || control.Handler.NativeView is null)
        {
            // Hack: When this executes the Handler and NativeView is null
            control.HandlerChanged += (_, __) =>
            {
                OnTintColorChanged(control, oldValue, newValue);
            };
            return;
        }
```
When the `PropertyChanged` is triggered the `Handler` will be `null`, so I need that hack to make it work. Here I don't have this property in all my `Image` controls, because I didn't add it to my `ImageHandler.Mapper`, but this isn't very well supported on the new architecture.

## [Platform Behavior API](https://github.com/pictos/MauiSpecRepo/tree/master/MauiSpecRepo/Behavior)

Using a specific API that can be attached to a particular control in the View will solve the issues with the `Mapper` approach and (with some implementation or changes) will prevent having the `Handler` and/or `MauiContext` as `null` values. The usage [will be](https://github.com/pictos/MauiSpecRepo/blob/master/MauiSpecRepo/MainPage.xaml#L14-L19)

```xml
 <Image x:Name="img"
        Source="shield.png">
     <Image.Behaviors>
         <local:TintColorBehavior TintColor="Red" />
     </Image.Behaviors>
 </Image>
```
As you can see nothing changed from the traditional `Behavior` usage, so devs don't need to learn how to use a new API.

In more complex scenarios, when we need to manage more properties and behaviors this approach is _better_ than the previous one mentioned here.

### Public API Changes

The API spec [will be](https://github.com/pictos/MauiSpecRepo/blob/master/MauiSpecRepo/BehaviorPlatform.cs)
> The API on my Github repo is different because today we've to work around the issue with the 
`Handler` being `null` when the `View` isn't.

```csharp
#if __IOS__
using NativeView = UIKit.UIView;
#elif __MACOS__
using NativeView = AppKit.NSView;
#elif ANDROID
using NativeView = Android.Views.View;
#elif WINDOWS
using NativeView = Microsoft.UI.Xaml.FrameworkElement;
#endif

namespace MauiSpecRepo;
/// <summary>
/// 
/// </summary>
public abstract partial class BasePlatformBehavior<TView, TNativeView> : Behavior<TView>
    where TView : VisualElement
#if !NET6_0 || IOS || ANDROID || WINDOWS
        where TNativeView : NativeView
#else
		where TNativeView : class
#endif
{
    protected TNativeView? NativeView => View?.Handler?.NativeView as TNativeView;
    protected TView? View { get; private set; }

    protected override void OnAttachedTo(TView bindable)
    {
        base.OnAttachedTo(bindable);
        View = bindable;
        OnPlatformAttachedBehavior(View);
    }

    protected override void OnDetachingFrom(TView bindable)
    {
        base.OnDetachingFrom(bindable);
        OnPlatformDeattachedBehavior(bindable);
    }

    protected abstract void OnPlatformAttachedBehavior(TView view);

    protected abstract void OnPlatformDeattachedBehavior(TView view);
}
```

With this implementation devs have:

- The feature (in this case the Tint color) added to a particular control;
- No need to add a new key on the `Mapper` and that be listened to by all `Image` controls;
- Intuitive places to clean up resources;

Changes in the framework to support this:
- Call this API when the `Handler` and `MauiContext` aren't `null`;
- Call the `OnDeatachingFrom` methods on the `DisconnectHandler` method;

### Intended Use-Case

This API be very useful not only for mobile developers but for library maintainers. This can replace all Effects API in the near future and use all the advantages of the language and multi-target arch.

I know that the .NET MAUI team is running against time, so if this is approved I can implement it and open a PR, if that helps you.

Cheers.

### PR DESCRIPTION
**Pull request by pictos**
### Description of Change

This PR implements a new API for Behaviors, that allows devs to make customizations in the platform-specific layer.
- Make it so Loaded will fire on a view if a user subscribes and Loaded has already fired
- If the Element the behavior is attached to is a VisualElement then we fire `OnAttached`/`OnDetached` on the `Loaded` event.
- If the Element doesn't have a `Loaded` event then we just fire when the `Handler` is added and removed

```C#
public abstract partial class PlatformBehavior<TView, TPlatformView> : Behavior<TView>
	where TView : Element
	where TPlatformView : class
{
	protected virtual void OnAttachedTo(TView bindable, TPlatformView platformView) { }
	protected virtual void OnDetachedFrom(TView bindable, TPlatformView platformView) { }
```

```C#
#if IOS
using PlatformView = UIKit.UIView;
#elif MACOS
using PlatformView = AppKit.NSView;
#elif ANDROID
using PlatformView = Android.Views.View;
#elif WINDOWS
using PlatformView = Microsoft.UI.Xaml.FrameworkElement;
#elif NET6_0 || NETSTANDARD
using PlatformView = System.Object;
#endif

public abstract partial class PlatformBehavior<TView> : PlatformBehavior<TView, PlatformView>
	where TView : VisualElement
{

}
```

### Example Usage

```C#
public class ChangeTextColor : PlatformBehavior<Label>
{
#if ANDROID
     protected override void OnAttachedTo(Label bindable, Android.Views.View platformView) 
     { 
     }
#endif
}
```

```XAML
<Label Text="Hello Behavior">
     <Label.Behaviors>
               <behavior:ChangeTextColor />
     </Label.Behaviors>
</Label>
```

### Issues 

Fixes #4958

### CODE REVIEW

**Review by pictos — COMMENTED** [review]
Hey, @Redth and @PureWeen here are the first versions of the API. I'll implement unit tests and samples after we ok with the implementation

File: src/Controls/src/Core/HandlerImpl/Element/Element.Impl.cs [hunk 1]
```diff
@@ -33,12 +33,29 @@ private protected virtual void OnHandlerChangedCore()
 			HandlerChanged?.Invoke(this, EventArgs.Empty);
 
 			OnHandlerChanged();
+			
+			if (this is VisualElement visual && visual.Behaviors.Count > 0 && Handler != null)
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Controls/src/Core/HandlerImpl/Element/Element.Impl.cs:37]: I feel like we should just use the `HandlerChanged` event inside the `Behavior` . Is there any reason not to?

[the remark this answers is not in the document]
Comment (pictos) [reply src/Controls/src/Core/HandlerImpl/Element/Element.Impl.cs:37]: I'm ok using it. @Redth are you ok with that? I talked to use the method that fires the event, but not sure if you thought in another kind of implementation

**Comment by PureWeen** [comment]
@pictos @Redth  I reworked this a bit 

- If the View Type is `VisualElement` then it uses `Loaded` for Attached/DetachedFrom. That just felt more consistent with the expectations of `Behavior`. They will only fire when `Loaded` fires
- If the View Type is `Element` then it uses `HandlerChanged`. It seemed useful to make it so `Behaviors` can work on `Elements` like `Window`

**Comment by pictos** [comment]
@PureWeen we still need to implement the `Unloaded` part, which will trigger the `OnDetachedFrom` method. Today Behaviors don't do that but talking with @Redth we thought that was a good idea to implement this here. What do you think? I can do the implementation Wednesday next week

**Comment by PureWeen** [comment]
@pictos  that's implemented here

https://github.com/dotnet/maui/pull/5043/files#diff-b47d1de690c9b627046ee616dc30e5be8f0ffbe906da8d2bfcc2400d448444f3R135

If the handler is removed or the the platform control is removed the platform visual tree then detached is called

**Review by PureWeen — APPROVED** [review]

**Pull request merged**